PUBDEV-5025: Add print method for AutoML and fill Stacked Ensemble when retrieved using getModel

tomasfryda opened the pull request:
https://0xdata.atlassian.net/browse/PUBDEV-5025

This PR consists of AutoML show/summary method and a fix for SE to fill in `model$model_summary` even when retrieved through `h2o.getModel()` so we can print at least something in the AutoML.  (Generic model uses that to print some basic statistics, more will be done in https://0xdata.atlassian.net/browse/PUBDEV-7315)

AutoML print (show method):
- project name
- leader model
- number of trained models
- start time
- end time
- duration
- leaderboard
- leader model details

AutoML summary - for now it's same as print with 2 exceptions:
- leaderboard shows all models
- leaderboard with all extra columns

Example of AutoML summary:
```
AutoML Details
==============
Project Name:  r_aml_get_automl 
Leader Model:  StackedEnsemble_AllModels_AutoML_20200608_161050 

Number of Trained Models:  7 
Start Time:  2020-06-08 16:10:50 
End Time:  2020-06-08 16:10:54 
Duration:  4 s

Leaderboard
===========
                                          model_id       auc   logloss
1 StackedEnsemble_AllModels_AutoML_20200608_161050 0.7962915 0.5357162
2                 XGBoost_2_AutoML_20200608_161050 0.7940745 0.5395687
3                 XGBoost_1_AutoML_20200608_161050 0.7915407 0.5597579
      aucpr mean_per_class_error      rmse       mse training_time_ms
1 0.6934854            0.2508710 0.4216215 0.1777647              115
2 0.7013956            0.2424923 0.4241721 0.1799220               47
3 0.6797751            0.2644467 0.4298898 0.1848052               37
  predict_time_per_row_ms
1                0.014533
2                0.006289
3                0.005393

[4 rows x 9 columns] 


Leader Model Details
====================
Model Details:
==============

H2OBinomialModel: stackedensemble
Model Key:  StackedEnsemble_AllModels_AutoML_20200608_161050 
Number of Base Models: 3

Base Models (count by algorithm type):

xgboost 
      3 

Metalearner:
.... (the rest of leader model summary)
```

sebhrusen commented on h2o-r/h2o-package/R/classes.R:
```diff
@@ -913,3 +913,57 @@ setClass("H2OAutoML", slots = c(project_name = "character",
+  summary(object@leader)
+
+  invisible(NULL)
+})
```
careful with EOF not at new line, some tools, doc generators for example, may be picky with that.

tomasfryda replied:
Thanks!  Fixed.

ledell commented:
Let's try to merge this PR first https://github.com/h2oai/h2o-3/pull/4369/commits

ledell commented:
@tomasfryda Any update on the licensing issue blocking the other (contributed) PR?  If that's not going to make it in, we should probably just merge this.

ledell commented on h2o-r/h2o-package/R/classes.R:
```diff
@@ -913,3 +913,57 @@ setClass("H2OAutoML", slots = c(project_name = "character",
+
+  cat("Leaderboard\n")
+  cat("===========\n")
+  print(h2o.get_leaderboard(object, "ALL"), n = -1)
```
@tomasfryda Is the `-1` a typo?  Why do you want to show all but the last row in the leaderboard?  I thought this was meant to show all the rows.

tomasfryda replied:
Thank you! You are absolutely right. Switching between java and R, I didn't realise that this omits the last row 🤦‍♂️

ledell commented:
@tomasfryda Your example above says that there were 7 models trained, but the leaderboard shows 3 rows.  Is that a bug?

tomasfryda commented:
@ledell Sorry, I think I edited the example to keep it shorter by removing lines from leaderboard. I tested it and it reports and shows the correct number of models.

About the contributed PR:
Unfortunately, I still don't have access to snyk to find out what is actually the problem... Github should look at license file[1] which wasn't modified. I will ask again on slack.

[1] https://help.github.com/en/github/creating-cloning-and-archiving-repositories/licensing-a-repository

tomasfryda commented:
@ledell I rebased on new master that includes Daniela's PR. So now it looks as follow:


```
> aml
AutoML Details
==============
Project Name:  AutoML_20200623_172110805 
Leader Model ID:  GLM_1_AutoML_20200623_172110 
Algorithm:  glm 

Number of Trained Models:  7 
Start Time:  2020-06-23 17:21:11 
End Time:  2020-06-23 17:21:18 
Duration:  7 s

Leaderboard
===========
                                             model_id mean_per_class_error   logloss      rmse
1                        GLM_1_AutoML_20200623_172110           0.05773420 0.1035326 0.1877265
2    StackedEnsemble_AllModels_AutoML_20200623_172110           0.07625272 0.3131982 0.2859207
3 StackedEnsemble_BestOfFamily_AutoML_20200623_172110           0.07625272 0.3146651 0.2860561
4                    XGBoost_3_AutoML_20200623_172110           0.07679739 0.2201469 0.2477941
5                        DRF_1_AutoML_20200623_172110           0.08551198 0.1898239 0.2482490
6                    XGBoost_1_AutoML_20200623_172110           0.08605664 0.3112504 0.2907549
7                    XGBoost_2_AutoML_20200623_172110           0.14433551 0.6273455 0.4641652
         mse
1 0.03524123
2 0.08175064
3 0.08182809
4 0.06140193
5 0.06162754
6 0.08453844
7 0.21544936

[7 rows x 5 columns] 


Leader Model Details
====================
Model Details:
==============

H2OMultinomialModel: glm
Model ID:  GLM_1_AutoML_20200623_172110 
GLM Model: summary
       family        link              regularization
1 multinomial multinomial Ridge ( lambda = 1.762E-4 )
                                                                    lambda_search
1 nlambda = 30, lambda.max = 42.193, lambda.min = 1.762E-4, lambda.1se = 0.001184
  number_of_predictors_total number_of_active_predictors number_of_iterations
1                         15                          12                  113
                   training_frame
1 automl_training_RTMP_sid_87ab_3

Coefficients: glm multinomial coefficients
      names coefs_class_0 coefs_class_1 coefs_class_2 std_coefs_class_0 std_coefs_class_1
1 Intercept      9.609389     -2.436025    -28.504866         -5.750489          0.194850
2 sepal_len     -2.226679      1.773052      0.763671         -1.776517          1.414598
3 sepal_wid      5.016029     -0.406294     -4.594725          2.097296         -0.169879
4 petal_len     -2.676156     -1.052591      4.279745         -4.617490         -1.816161
5 petal_wid     -5.609740     -1.957516      8.919997         -4.204996         -1.467331
  std_coefs_class_2
1        -10.067141
2          0.609282
3         -1.921140
4          7.384354
5          6.686327

H2OMultinomialMetrics: glm
** Reported on training data. **

Training Set Metrics: 
=====================

Extract training frame with `h2o.getFrame("automl_training_RTMP_sid_87ab_3")`
MSE: (Extract with `h2o.mse`) 0.01902078
RMSE: (Extract with `h2o.rmse`) 0.1379158
Logloss: (Extract with `h2o.logloss`) 0.06237409
Mean Per-Class Error: 0.01906318
Null Deviance: (Extract with `h2o.nulldeviance`) 219.1563
Residual Deviance: (Extract with `h2o.residual_deviance`) 12.47482
R^2: (Extract with `h2o.r2`) 0.9710226
AIC: (Extract with `h2o.aic`) NaN
Confusion Matrix: Extract with `h2o.confusionMatrix(<model>,train = TRUE)`)
=========================================================================
Confusion Matrix: Row labels: Actual class; Column labels: Predicted class
                Iris-setosa Iris-versicolor Iris-virginica  Error      Rate
Iris-setosa              30               0              0 0.0000 =  0 / 30
Iris-versicolor           0              33              1 0.0294 =  1 / 34
Iris-virginica            0               1             35 0.0278 =  1 / 36
Totals                   30              34             36 0.0200 = 2 / 100

Hit Ratio Table: Extract with `h2o.hit_ratio_table(<model>,train = TRUE)`
=======================================================================
Top-3 Hit Ratios: 
  k hit_ratio
1 1  0.980000
2 2  1.000000
3 3  1.000000



H2OMultinomialMetrics: glm
** Reported on cross-validation data. **
** 5-fold cross-validation on training data (Metrics computed for combined holdout predictions) **

Cross-Validation Set Metrics: 
=====================

Extract cross-validation frame with `h2o.getFrame("automl_training_RTMP_sid_87ab_3")`
MSE: (Extract with `h2o.mse`) 0.03524123
RMSE: (Extract with `h2o.rmse`) 0.1877265
Logloss: (Extract with `h2o.logloss`) 0.1035326
Mean Per-Class Error: 0.0577342
Null Deviance: (Extract with `h2o.nulldeviance`) 222.1446
Residual Deviance: (Extract with `h2o.residual_deviance`) 20.70652
R^2: (Extract with `h2o.r2`) 0.9463114
AIC: (Extract with `h2o.aic`) NaN
Hit Ratio Table: Extract with `h2o.hit_ratio_table(<model>,xval = TRUE)`
=======================================================================
Top-3 Hit Ratios: 
  k hit_ratio
1 1  0.940000
2 2  1.000000
3 3  1.000000


Cross-Validation Metrics Summary: 
                               mean          sd  cv_1_valid cv_2_valid  cv_3_valid  cv_4_valid
accuracy                       0.94 0.041833002         1.0        0.9         0.9        0.95
err                            0.06 0.041833002         0.0        0.1         0.1        0.05
err_count                       1.2     0.83666         0.0        2.0         2.0         1.0
logloss                  0.09338274  0.04609131  0.03870104 0.11480715  0.15473911  0.10041821
max_per_class_error      0.14801587  0.09794436         0.0       0.25  0.22222222       0.125
mean_per_class_accuracy  0.94232804  0.04605806         1.0      0.875   0.9259259   0.9583333
mean_per_class_error    0.057671957  0.04605806         0.0      0.125 0.074074075 0.041666668
mse                      0.03069646 0.016494129 0.009982957 0.04089475  0.05184705 0.030441754
null_deviance              44.42893   0.6722112    44.91473  45.299534   44.338173   43.842693
r2                       0.95171297 0.028944384  0.98307973 0.94888157  0.90530217  0.95588154
residual_deviance         3.7353094   1.8436524   1.5480416   4.592286   6.1895647   4.0167284
rmse                     0.16936956  0.05012995 0.099914744 0.20222451  0.22769947  0.17447565
                         cv_5_valid
accuracy                       0.95
err                            0.05
err_count                       1.0
logloss                  0.05824816
max_per_class_error      0.14285715
mean_per_class_accuracy  0.95238096
mean_per_class_error     0.04761905
mse                     0.020315783
null_deviance              43.74951
r2                       0.96541995
residual_deviance         2.3299263
rmse                     0.14253345
```
```
> summary(aml)
AutoML Summary
==============
Project Name:  AutoML_20200623_172110805 
Leader Model ID:  GLM_1_AutoML_20200623_172110 
Algorithm:  glm 

Number of Trained Models:  7 
Start Time:  2020-06-23 17:21:11 
End Time:  2020-06-23 17:21:18 
Duration:  7 s

Leaderboard
===========
                                             model_id mean_per_class_error   logloss      rmse
1                        GLM_1_AutoML_20200623_172110           0.05773420 0.1035326 0.1877265
2    StackedEnsemble_AllModels_AutoML_20200623_172110           0.07625272 0.3131982 0.2859207
3 StackedEnsemble_BestOfFamily_AutoML_20200623_172110           0.07625272 0.3146651 0.2860561
4                    XGBoost_3_AutoML_20200623_172110           0.07679739 0.2201469 0.2477941
5                        DRF_1_AutoML_20200623_172110           0.08551198 0.1898239 0.2482490
6                    XGBoost_1_AutoML_20200623_172110           0.08605664 0.3112504 0.2907549
7                    XGBoost_2_AutoML_20200623_172110           0.14433551 0.6273455 0.4641652
         mse training_time_ms predict_time_per_row_ms
1 0.03524123              163                0.009153
2 0.08175064              222                0.056971
3 0.08182809              120                0.034371
4 0.06140193               30                0.005100
5 0.06162754              101                0.014284
6 0.08453844               99                0.004979
7 0.21544936               38                0.004917

[7 rows x 7 columns] 


Leader Model Details
====================
Model Details:
==============

H2OMultinomialModel: glm
Model Key:  GLM_1_AutoML_20200623_172110 
GLM Model: summary
       family        link              regularization
1 multinomial multinomial Ridge ( lambda = 1.762E-4 )
                                                                    lambda_search
1 nlambda = 30, lambda.max = 42.193, lambda.min = 1.762E-4, lambda.1se = 0.001184
  number_of_predictors_total number_of_active_predictors number_of_iterations
1                         15                          12                  113
                   training_frame
1 automl_training_RTMP_sid_87ab_3

H2OMultinomialMetrics: glm
** Reported on training data. **

Training Set Metrics: 
=====================

Extract training frame with `h2o.getFrame("automl_training_RTMP_sid_87ab_3")`
MSE: (Extract with `h2o.mse`) 0.01902078
RMSE: (Extract with `h2o.rmse`) 0.1379158
Logloss: (Extract with `h2o.logloss`) 0.06237409
Mean Per-Class Error: 0.01906318
Null Deviance: (Extract with `h2o.nulldeviance`) 219.1563
Residual Deviance: (Extract with `h2o.residual_deviance`) 12.47482
R^2: (Extract with `h2o.r2`) 0.9710226
AIC: (Extract with `h2o.aic`) NaN
Confusion Matrix: Extract with `h2o.confusionMatrix(<model>,train = TRUE)`)
=========================================================================
Confusion Matrix: Row labels: Actual class; Column labels: Predicted class
                Iris-setosa Iris-versicolor Iris-virginica  Error      Rate
Iris-setosa              30               0              0 0.0000 =  0 / 30
Iris-versicolor           0              33              1 0.0294 =  1 / 34
Iris-virginica            0               1             35 0.0278 =  1 / 36
Totals                   30              34             36 0.0200 = 2 / 100

Hit Ratio Table: Extract with `h2o.hit_ratio_table(<model>,train = TRUE)`
=======================================================================
Top-3 Hit Ratios: 
  k hit_ratio
1 1  0.980000
2 2  1.000000
3 3  1.000000



H2OMultinomialMetrics: glm
** Reported on cross-validation data. **
** 5-fold cross-validation on training data (Metrics computed for combined holdout predictions) **

Cross-Validation Set Metrics: 
=====================

Extract cross-validation frame with `h2o.getFrame("automl_training_RTMP_sid_87ab_3")`
MSE: (Extract with `h2o.mse`) 0.03524123
RMSE: (Extract with `h2o.rmse`) 0.1877265
Logloss: (Extract with `h2o.logloss`) 0.1035326
Mean Per-Class Error: 0.0577342
Null Deviance: (Extract with `h2o.nulldeviance`) 222.1446
Residual Deviance: (Extract with `h2o.residual_deviance`) 20.70652
R^2: (Extract with `h2o.r2`) 0.9463114
AIC: (Extract with `h2o.aic`) NaN
Hit Ratio Table: Extract with `h2o.hit_ratio_table(<model>,xval = TRUE)`
=======================================================================
Top-3 Hit Ratios: 
  k hit_ratio
1 1  0.940000
2 2  1.000000
3 3  1.000000


Cross-Validation Metrics Summary: 
                               mean          sd  cv_1_valid cv_2_valid  cv_3_valid  cv_4_valid
accuracy                       0.94 0.041833002         1.0        0.9         0.9        0.95
err                            0.06 0.041833002         0.0        0.1         0.1        0.05
err_count                       1.2     0.83666         0.0        2.0         2.0         1.0
logloss                  0.09338274  0.04609131  0.03870104 0.11480715  0.15473911  0.10041821
max_per_class_error      0.14801587  0.09794436         0.0       0.25  0.22222222       0.125
mean_per_class_accuracy  0.94232804  0.04605806         1.0      0.875   0.9259259   0.9583333
mean_per_class_error    0.057671957  0.04605806         0.0      0.125 0.074074075 0.041666668
mse                      0.03069646 0.016494129 0.009982957 0.04089475  0.05184705 0.030441754
null_deviance              44.42893   0.6722112    44.91473  45.299534   44.338173   43.842693
r2                       0.95171297 0.028944384  0.98307973 0.94888157  0.90530217  0.95588154
residual_deviance         3.7353094   1.8436524   1.5480416   4.592286   6.1895647   4.0167284
rmse                     0.16936956  0.05012995 0.099914744 0.20222451  0.22769947  0.17447565
                         cv_5_valid
accuracy                       0.95
err                            0.05
err_count                       1.0
logloss                  0.05824816
max_per_class_error      0.14285715
mean_per_class_accuracy  0.95238096
mean_per_class_error     0.04761905
mse                     0.020315783
null_deviance              43.74951
r2                       0.96541995
residual_deviance         2.3299263
rmse                     0.14253345

Scoring History: 
            timestamp   duration iteration lambda predictors deviance_train deviance_test
1 2020-06-23 17:21:14  0.000 sec         2  .42E2         15          2.150            NA
2 2020-06-23 17:21:14  0.005 sec         4  .26E2         15          2.126            NA
3 2020-06-23 17:21:14  0.009 sec         6  .16E2         15          2.089            NA
4 2020-06-23 17:21:14  0.012 sec         8   .1E2         15          2.035            NA
5 2020-06-23 17:21:14  0.016 sec        10  .63E1         15          1.956            NA
  deviance_xval deviance_se
1         2.188       0.014
2         2.168       0.013
3         2.138       0.012
4         2.092       0.011
5         2.025       0.010

---
             timestamp   duration iteration lambda predictors deviance_train deviance_test
25 2020-06-23 17:21:14  0.129 sec        94 .46E-3         15          0.140            NA
26 2020-06-23 17:21:14  0.135 sec       100 .28E-3         15          0.131            NA
27 2020-06-23 17:21:14  0.141 sec       105 .18E-3         15          0.125            NA
28 2020-06-23 17:21:14  0.146 sec       108 .11E-3         15          0.120            NA
29 2020-06-23 17:21:14  0.150 sec       111 .68E-4         15          0.117            NA
30 2020-06-23 17:21:14  0.154 sec       113 .42E-4         15          0.115            NA
   deviance_xval deviance_se
25         0.214       0.027
26         0.209       0.032
27         0.207       0.037
28         0.209       0.044
29         0.215       0.051
30         0.000       0.000

Variable Importances: (Extract with `h2o.varimp`) 
=================================================

   variable relative_importance scaled_importance percentage
1 petal_len           13.818005         1.0000000  0.4044447
2 petal_wid           12.358655         0.8943878  0.3617304
3 sepal_wid            4.188315         0.3031056  0.1225895
4 sepal_len            3.800396         0.2750322  0.1112353
```

ledell commented:
@tomasfryda I think the text of the print and summary methods are too long in the case where a regular (non-stacked ensemble model) is the leader.   I think the better long-term solution is to make shorter summary methods for H2O models (by removing the long model metrics print-outs) a keep the print-out of the leader model at the end, but in the mean-time, I think we should just remove the print-out of the leader model object since the text is so long.

Also, can you make the following changes with this text?

- Change to "Total Number of Models Trained":
- Add timezone to start and end time

```
Number of Trained Models:  7 
Start Time:  2020-06-23 17:21:11 
End Time:  2020-06-23 17:21:18 
Duration:  7 s
```

ledell commented:
Hey @tomasfryda When you have a chance, can you delete the print-out of the leader model at the end?  That's the only change for now.  (We can consider adding more info later, but for now this is a big improvement). 

I was also wondering if we can make this PR to rel-zahradnik instead?  I'd love to get this into the next CRAN release (3.30.0.6).  Let me know if that causes issues with the Daniele's PR or if you see any other issues with pushing this to the currently release vs master.

tomasfryda commented:
Hey @ledell, sorry I forgot to mention it during the call. I already removed the leader model printout and added timezone using `h2o.getTimezone()`. About `rel-zahradnik`, Daniele's PR was merged to `master` so I guess that could cause conflicts during automatic merge.

ledell commented:
@tomasfryda Thanks!  Ok, then let's just merge to master.
